libs/langchain-community/src/vectorstores/upstash.ts

```diff
@@ -0,0 +1,241 @@
+  index: UpstashIndex;
+
+  caller: AsyncCaller;
+
+  _vectorstoreType(): string {
+    return "upstash";
+  }
+
+  constructor(embeddings: EmbeddingsInterface, args: UpstashVectorLibArgs) {
+    super(embeddings, args);
+
+    this.embeddings = embeddings;
+
+    const { index, ...asyncCallerArgs } = args;
+
+    this.index = index;
+    this.caller = new AsyncCaller(asyncCallerArgs);
+  }
+
+  /**
+   * This method adds documents to Upstash database. Documents are first converted to vectors
+   * using the provided embeddings instance, and then upserted to the database.
+   * @param documents Array of Document objects to be added to the database.
+   * @param options Optional object containing array of ids for the documents.
+   * @returns Promise that resolves with the ids of the provided documents when the upsert operation is done.
+   */
+  async addDocuments(documents: Document[], options?: { ids?: string[] }) {
+    const texts = documents.map(({ pageContent }) => pageContent);
+
+    const embeddings = await this.embeddings.embedDocuments(texts);
+
+    return this.addVectors(embeddings, documents, options);
+  }
+
+  /**
+   * This method adds the provided vectors to Upstash database.
+   * @param vectors  Array of vectors to be added to the Upstash database.
+   * @param documents Array of Document objects, each associated with a vector.
+   * @param options Optional object containing the array of ids foor the vectors.
+   * @returns Promise that resolves with the ids of the provided documents when the upsert operation is done.
+   */
+  async addVectors(
+    vectors: number[][],
+    documents: Document[],
+    options?: { ids?: string[] }
+  ) {
+    const documentIds =
+      options?.ids ?? Array.from({ length: vectors.length }, () => uuid.v4());
+
+    const upstashVectors = vectors.map((vector, index) => {
+      const metadata = {
+        pageContentLC: documents[index].pageContent,
+        ...documents[index].metadata,
+      };
+
+      const id = documentIds[index];
+
+      return {
+        id,
+        vector,
+        metadata,
+      };
+    });
+
+    const vectorChunks = chunkArray(upstashVectors, CONCURRENT_UPSERT_LIMIT);
+
+    const batchRequests = vectorChunks.map((chunk) =>
+      this.caller.call(async () => this.index.upsert(chunk))
+    );
+
+    await Promise.all(batchRequests);
+
+    return documentIds;
+  }
+
+  /**
+   * This method deletes documents from the Upstash database. You can either
+   * provide the target ids, or delete all vectors in the database.
+   * @param params Object containing either array of ids of the documents or boolean deleteAll.
+   * @returns Promise that resolves when the specified documents have been deleted from the database.
+   */
+  async delete(params: UpstashDeleteParams): Promise<void> {
+    if (params.deleteAll) {
+      await this.index.reset();
+    } else if (params.ids) {
+      await this.index.delete(params.ids);
+    } else {
```

Comment:
I'm pretty sure TS is smart enough to recognize the `never` and not need this `else`, could you double check we don't need this?